### PR TITLE
Prepare Asciidoctor 2.0.0

### PR DESCRIPTION
**Pull request by Mogztter**
This pull request is compatible with: https://github.com/asciidoctor/asciidoctor/pull/2946

```
ASCIIDOCTOR_CORE_VERSION=mojavelinux/asciidoctor#issue-2764-no-unmaintained-rubies npm run build:quick
```

### CODE REVIEW

**Comment by mojavelinux** [comment]
Actually, I might put back that `require 'logger'` statement since it doesn't follow the scenario of the other autoloads. We always require logging, so there's no benefit to using autoload here.

**Comment by mojavelinux** [comment]
We no longer use the Kernel.open function, so although it doesn't hurt to keep it, that could be removed.

**Comment by mojavelinux** [comment]
We no longer use the Kernel.open function, so although it doesn't hurt to keep it, that could be removed.

**Comment by mojavelinux** [comment]
The only modifications that seem to be needed are:

* Mapping File.absolute_path to File.expand_path
* Mapping Kernel.__dir__

The only time either of these should be used by Asciidoctor.js is when defining the ROOT_DIR constant. I wonder if that constant should be defined by the asciidoctor/js require. Then, we could avoid this whole code path.

**Comment by mojavelinux** [comment]
I'm mulling over the constant names FILE_READ_MODE and URI_READ_MODE since they become part of the integration API. Ruby calls these "open modes". However, they aren't just any open mode, but specifically for reads. Do you think these names are clear, or should we name them something like FILE_READ_OPEN_MODE? (I might be overthinking this).

**Comment by Mogztter** [comment]
> The only time either of these should be used by Asciidoctor.js is when defining the ROOT_DIR constant. I wonder if that constant should be defined by the asciidoctor/js require. Then, we could avoid this whole code path.

Yes I think it's a good idea to define this constant in `asciidoctor/js` require.
We should also do the same with `STYLESHEETS_DATA_PATH` (instead of patching the code with a regexp).

> I'm mulling over the constant names FILE_READ_MODE and URI_READ_MODE since they become part of the integration API. Ruby calls these "open modes". However, they aren't just any open mode, but specifically for reads. Do you think these names are clear, or should we name them something like FILE_READ_OPEN_MODE? (I might be overthinking this).

In my opinion, `FILE_READ_MODE` and `URI_READ_MODE` are fine :grinning:

**Comment by mojavelinux** [comment]
> Yes I think it's a good idea to define this constant in asciidoctor/js require.

I suppose we could add:

```
ROOT_DIR = ::File.dirname ::File.absolute_path __dir__ unless defined? ROOT_DIR
```

wdyt?

> In my opinion, FILE_READ_MODE and URI_READ_MODE are fine grinning

:+1:

**Comment by mojavelinux** [comment]
For stylesheets, I think it's more correct to override the methods. The constants in that class should be considered internal. You don't have to satisfy them. You just have to satisfy the Stylesheets API. In fact, we could even stop loading it in Opal (so you have total control).

File: lib/asciidoctor/js.rb [hunk 1]
```diff
@@ -1,3 +1,4 @@
 require 'asciidoctor/js/opal_ext'
 require 'asciidoctor/js/rx'
 require 'strscan'
+require 'logger'
```
Comment (mojavelinux) [inline lib/asciidoctor/js.rb]: This is now loaded by default again.

File: lib/asciidoctor/js/opal_ext/kernel.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module Kernel
+  # basic implementation of open, enough to work with reading files over XMLHttpRequest
+  def open(path, *rest)
```
Comment (mojavelinux) [inline lib/asciidoctor/js/opal_ext/kernel.rb]: We'll no longer be using Kernel.open in Asciidoctor core starting in 2.0.

**Comment by mojavelinux** [comment]
> Then I could override the method `primary_stylesheet_path` to resolve the path according to the JavaScript environment.

I'd even go up one level. How about `stylesheet_data_path`? That way, we can use that method to get any stylesheet we happen to bundle.

> I'm currently debugging the bom test failure. 

I may have made a mistake here. I'm looking into it.

**Comment by mojavelinux** [comment]
I was using slice instead of byteslice. slice relies on how bytes are interpreted in the encoding. What we want is byteslice, which just rips out bytes 1-by-1.

**Comment by mojavelinux** [comment]
PR updated.

**Comment by Mogztter** [comment]
Build is green, should I move forward and merge this pull request before asciidoctor/asciidoctor#2946 ?

**Comment by mojavelinux** [comment]
Yep! I'm about to merge that PR.

**Pull request merged**